keras/backend/theano_backend.py

```diff
@@ -1551,9 +1551,13 @@ def relu(x, alpha=0., max_value=None):
     return x
 
 
-def softmax(x):
-    return T.nnet.softmax(x)
-
+def softmax(x, axis=-1):
```

Comment:
@fchollet This is what I could think of. Using transpose causes a "theano expects 1D/2D array, input is 3D" error for 3 dimensions.
This method gives 77ms mean time. The earlier one gave 60ms.

Reply:
Oh, you didn't have to actually implement the general case! Benchmarking ndim=2 and axis=0 would have been fine (and wouldn't require collapsing the array into 2D).

If the "manual" softmax computation is faster than the reshape/transpose, then let's go with that.

Thanks!

Reply:
Does CNTK support a softmax axis? If not, we can also go with manual softmax in CNTK.

Reply:
I think the manual one is faster for some reason (atleast in the general n-dimensional case that I benchmarked anyway), so reverted to that.
CNTK does support axis for softmax: https://www.cntk.ai/pythondocs/cntk.ops.html?highlight=softmax#cntk.ops.softmax